dds/idl/marshal_generator.cpp

```diff
@@ -909,6 +909,422 @@ namespace {
+              intro.join(be_global->impl_, indent);
+              be_global->impl_ <<
+                indent << "serialized_size(encoding, size, " << elem_wrapper.ref() << ");\n";
+            }
+          }
+
+          if (!val_generated) { // Value
+            // String, Struct, Array, Sequence, Union
+            if (val_cls & CL_STRING) {
+              be_global->impl_ << indent <<
+                "primitive_serialized_size_ulong(encoding, size);\n";
+              const string strlen_suffix = (val_cls & CL_WIDE)
+                ? " * char16_cdr_size;\n" : " + 1;\n";
+              be_global->impl_ << indent <<
+                "size += it->second.size()" << strlen_suffix;
+            } else {
+              RefWrapper elem_wrapper(key, key_cxx_elem, "it->second");
+              elem_wrapper.nested_key_only_ = nested_key_only;
+              Intro intro;
+              elem_wrapper.done(&intro);
+              const std::string indent = "    ";
+              intro.join(be_global->impl_, indent);
+              be_global->impl_ <<
+                indent << "serialized_size(encoding, size, " << elem_wrapper.ref() << ");\n";
+            }
+          }
+
+          be_global->impl_ <<
+            "  }\n";
+        }
+      }
+    }
+
+    {
+      Intro intro;
+      RefWrapper wrapper(base_wrapper);
+      wrapper.done(&intro);
+      const std::string value_access = wrapper.value_access();
+      const std::string get_length = wrapper.map_get_length();
+      const std::string check_empty = wrapper.map_check_empty();
+      Function insertion("operator<<", "bool");
+      insertion.addArg("strm", "Serializer&");
+      insertion.addArg("map", wrapper.wrapped_type_name());
+      insertion.endArgs();
+
+      if ((key_cls & CL_INTERFACE) == 0) {
+        be_global->impl_ <<
+          "  const Encoding& encoding = strm.encoding();\n"
+          "  ACE_UNUSED_ARG(encoding);\n";
+        marshal_generator::generate_dheader_code(
+          "    serialized_size(encoding, total_size, map);\n"
+          "    if (!strm.write_delimiter(total_size)) {\n"
+          "      return false;\n"
+          "    }\n", !val_primitive);
+
+        intro.join(be_global->impl_, "  ");
+
+        be_global->impl_ <<
+          "  const CORBA::ULong length = " << get_length << ";\n";
+
+        be_global->impl_ <<
+          streamAndCheck("<< length") <<
+          "  if (length == 0) {\n"
+          "    return true;\n"
+          "  }\n";
+      }
+
+      be_global->impl_ <<
+        "  for (auto it = " << value_access <<  ".begin(); it != " << value_access << ".end(); ++it) {\n";
+
+      if (key_cls & CL_INTERFACE || val_cls & CL_INTERFACE) {
+        be_global->impl_ <<
+          "  return false; // map with either a key or value of objrefs is not marshaled\n";
+      } else if (key_cls & CL_UNKNOWN || val_cls & CL_UNKNOWN) {
+        be_global->impl_ <<
+          "  return false; // map with either key or value of unknown/unsupported type\n";
```

Comment:
It would probably be helpful for users to indicate if the key or the value type is the problem -- unless this is handled somewhere else?